addons/hr_expense/models/hr_expense.py

```diff
@@ -454,8 +458,28 @@ def write(self, vals):
                     self.sheet_id.write({'employee_id': vals['employee_id']})
                 elif len(employees) > 1:
                     self.sheet_id = False
+        if 'sheet_id' in vals:
+            for expense in self:
+                prev_sheet = expense_to_prev_sheet[expense]
+                checksums = set((expense.attachment_ids - prev_sheet.expense_line_ids.attachment_ids).mapped('checksum'))
+                prev_sheet.attachment_ids.filtered(lambda att: att.checksum in checksums).with_context(sync_attachment=False).unlink()
+                if vals['sheet_id']:
+                    if expense.sheet_id != prev_sheet:
```

Comment:
A reason why this isn't higher in the tree?
It feels like it could unlink to copy the same one in weird corner-cases

Reply:
>A reason why this isn't higher in the tree?

Since we fall in this case not only when we write with `sheet_id` on the expense itself, but also when processing commands on the `expense_line_ids` of the sheet itself, I needed to handle all this after the `super`.
I need to know what are the attachments that will be left on the previous sheet so I know what to sync.

>It feels like it could unlink to copy the same one in weird corner-cases

Normally the no-op case is covered correctly: for the deletion we substract the attachments on the expense lines of the previous sheet, this prevents deleting attachments when we write with the same sheet_id as the current expense.
For the copy, we just check that the new sheet of the expense is != than the old one, this will cover potential duplication.

If you see a corner case I don't see, let me know, it's a tricky segment :S